### PR TITLE
Remove one of the two worklists from the SIL linker

### PR DESCRIPTION
**Pull request by slavapestov**
I'm re-doing the control flow here in preparation for landing a change making the linker lazier in -Onone.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/SIL/Linker.h [hunk 1]
```diff
@@ -79,10 +79,9 @@ class SILLinkerVisitor : public SILInstructionVisitor<SILLinkerVisitor, bool> {
   bool visitMetatypeInst(MetatypeInst *MI);
 
 private:
-  /// Add a function to our function worklist for processing.
-  void addFunctionToWorklist(SILFunction *F) {
-    FunctionDeserializationWorklist.push_back(F);
-  }
+  bool addFunctionToWorklist(SILFunction *F);
+
+  bool maybeAddFunctionToWorklist(SILFunction *F);
```
Comment (gottesmm) [inline lib/SIL/Linker.h]: Please add doxygen comments here.

File: lib/SIL/Linker.cpp [hunk 2]
```diff
@@ -70,10 +85,7 @@ bool SILLinkerVisitor::linkInVTable(ClassDecl *D) {
   // for processing.
   bool Result = false;
   for (auto P : Vtbl->getEntries()) {
-    if (P.Implementation->isExternalDeclaration()) {
-      Result = true;
-      addFunctionToWorklist(P.Implementation);
-    }
+    Result |= maybeAddFunctionToWorklist(P.Implementation);
```
Comment (gottesmm) [inline lib/SIL/Linker.cpp]: I haven't thought about this (I am asking you to do so), but are you sure that having a uniform condition across all of these is safe. For instance just off the top of my head, imagine if we could specialize witness tables. Would the linkage be shared? If we then serialized that, wouldn't we have shared external linkage and then wouldn't deserialize the function?

Also given that you are changing the conditions, can you add tests that verify this new behavior for other parts of this patch?

File: lib/SIL/Linker.cpp [hunk 3]
```diff
@@ -218,7 +219,7 @@ bool SILLinkerVisitor::visitProtocolConformance(
       // to deserialize everything, add the function to the list of functions
       // to deserialize.
       performFuncDeserialization = true;
-      addFunctionToWorklist(E.getMethodWitness().Witness);
+      maybeAddFunctionToWorklist(E.getMethodWitness().Witness);
```
Comment (gottesmm) [inline lib/SIL/Linker.cpp]: You are also expanding the conditions here. Same concern: what about specialized witness tables? That being said, maybe this is not interesting.

**Review by gottesmm — COMMENTED** [review]
Reviewed the first patch. I need to look at the other ones in context of the branch to understand.

File: lib/SIL/Linker.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SIL/Linker.cpp]: Safe is the wrong word here, I meant are you sure that expanding the conditions here what we want to do.

Comment (slavapestov) [reply lib/SIL/Linker.cpp]: hasSharedVisibility() returns true for both shared and shared external. Also note that linkInVTable() is only called it isLinkAll() is true, so in this case the condition is equivalent to F->isExternalDefinition(). So there’s no real functional change here. I’ll add an assert and comments.

File: lib/SIL/Linker.cpp (hunk 3 above)
Comment (slavapestov) [reply lib/SIL/Linker.cpp]: The only way we can end here in the mandatory pipeline (where isLinkAll is off) is for imported conformances, where all witnesses are shared. So again there’s no real change here. We are just avoiding adding functions to the work list if they already have a body.

**Comment by slavapestov** [comment]
@gottesmm Should we rename LinkNormal to LinkMandstory and LinkAll to LinkPerformance?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**